scripts/feature_tests/upgrade/controlplane_components_upgrade/upgrade_CAPI_and_CAPM3_with_clusterctl.sh

```diff
@@ -0,0 +1,98 @@
+}
+EOF
+
+cat <<EOF >/home/"${USER}"/.cluster-api/clusterctl.yaml
+providers:
+  - name: cluster-api
+    url: /home/$USER/.cluster-api/overrides/cluster-api/v0.3.1/core-components.yaml
+    type: CoreProvider
+  - name: kubeadm
+    url: /home/$USER/.cluster-api/overrides/bootstrap-kubeadm/v0.3.2/bootstrap-components.yaml
+    type: BootstrapProvider
+  - name: kubeadm
+    url: /home/$USER/.cluster-api/overrides/control-plane-kubeadm/v0.3.3/control-plane-components.yaml
+    type: ControlPlaneProvider
+  - name: metal3
+    url: /home/$USER/.cluster-api/overrides/infrastructure-metal3/v0.3.4/infrastructure-components.yaml
+    type: InfrastructureProvider
+EOF
+mv clusterctl-settings-metal3.json "${CAPM3_REPO}/clusterctl-settings.json"
+
+# Install initial version
+clusterctl_init_command=$(cmd/clusterctl/hack/local-overrides.py | grep "clusterctl init")
+echo "${clusterctl_init_command}" --target-namespace t1 --watching-namespace t1 | bash
+
+# Create a new version
+cp -r /home/"${USER}"/.cluster-api/overrides/cluster-api/v0.3.0 /home/"${USER}"/.cluster-api/overrides/cluster-api/v0.3.1
+cp -r /home/"${USER}"/.cluster-api/overrides/bootstrap-kubeadm/v0.3.0 /home/"${USER}"/.cluster-api/overrides/bootstrap-kubeadm/v0.3.2
+cp -r /home/"${USER}"/.cluster-api/overrides/control-plane-kubeadm/v0.3.0 /home/"${USER}"/.cluster-api/overrides/control-plane-kubeadm/v0.3.3
+cp -r /home/"${USER}"/.cluster-api/overrides/infrastructure-metal3/v0.3.0 /home/"${USER}"/.cluster-api/overrides/infrastructure-metal3/v0.3.4
+
+popd || exit
+# Make changes on CRDs
+sed -i 's/\bma\b/ma2020/g' /home/"${USER}"/.cluster-api/overrides/cluster-api/v0.3.1/core-components.yaml
+sed -i 's/singular: kubeadmconfig/singular: kubeadmconfig2020/' /home/"${USER}"/.cluster-api/overrides/bootstrap-kubeadm/v0.3.2/bootstrap-components.yaml
+sed -i 's/kcp/kcp2020/' /home/"${USER}"/.cluster-api/overrides/control-plane-kubeadm/v0.3.3/control-plane-components.yaml
+sed -i 's/\bm3c\b/m3c2020/g' /home/"${USER}"/.cluster-api/overrides/infrastructure-metal3/v0.3.4/infrastructure-components.yaml
+
+# do upgrade
+clusterctl upgrade plan | grep "upgrade apply" | xargs | xargs clusterctl
+
+# Verify upgrade
+upgraded_controllers_count=$(kubectl api-resources | grep -c 2020)
+upgraded_bootstrap_crd_count=$(kubectl get crds kubeadmconfigs.bootstrap.cluster.x-k8s.io -o json | jq '.spec.names.singular' | wc -l)
+upgraded_capm3_controller_count=$(kubectl api-resources | grep -c m3c2020)
+
+if [ "${upgraded_controllers_count}" -ne 2 ]; then
+  echo "Failed to upgrade cluster-api and controlplane components"
+  log_test_result "${0}" "fail"
+  exit 1
+fi
+if [ "${upgraded_bootstrap_crd_count}" -ne 1 ]; then
+  echo "Failed to upgrade control-plane-kubeadm components"
+  log_test_result "${0}" "fail"
+  exit 1
+fi
+
+if [ "${upgraded_capm3_controller_count}" -ne 1 ]; then
+  echo "Failed to upgrad infrastructure components"
+  log_test_result "${0}" "fail"
+  exit 1
+fi
+
```

Comment:
you should also verify here that the controllers are properly running

Reply:
done